### ISSUE TITLE
Unable to connect lines which belong to one and the same relation in some cases

### ISSUE THREAD

**Issue by yaugenka** [openstreetmap/iD#4769]
The two lines belong two one and the same mutipolygon relation. Pay attention to the direction of the lines.
You can connect end with beginning but cannot connect beginning with end.
You can connect beginning with beginning but cannot conneect end with end.
The connection should be possible in all these cases.
![line_connect](https://user-images.githubusercontent.com/14937001/35770068-fe735aca-0925-11e8-8ce7-c7f435f7da1d.gif)

**Comment by bhousel** [comment]
Interesting, thanks for flagging this..
I'm not surprised that there are issues like this - the pictured lines do not form a valid multipolygon.

**Comment by bhousel** [comment]
The issue is related to these lines:

https://github.com/openstreetmap/iD/blob/35a7a4c3beb1161614a0f5c2676c73cd241285de/modules/modes/drag_node.js#L224-L229

It's supposed to skip over the checks if the parent is not a multipolygon.  In this case - the code _thinks_ it's a multipolygon, but it's not a very good multipolygon, and so the `osmJoinWays` stuff that happens later has no hope of working :)

**Comment by yaugenka** [comment]
Please note the case when a polygon or multipolygon is created manually out of several unclosed connected lines which is a common thing for big polygons like administrative boundaries.

**Comment by slhh** [comment]
Joining the rings seems to be questionable here. The rings might not be fully loaded anyway.

For dragging a node a more local check seems to be sufficient. The goal is to prevent breaking a multipolygon, but not to prevent modifying a broken multipolygon.

We need to prevent:
1. Any intersection of the active (moved) edges (excluding nodes) of the multipolygon with any other edge of the multpolygon, including intersections in the nodes of the other edge. Already existing intersection should be ignored, even if they are moved to the other active edge.
2. Any connection of the dragged node to any inactive edge of the multipolygon (excluding nodes)
3. Any connection of the dragged node to any other node of the multipolygon, with the following exceptions:
    - Connection to the other node of an active (moved) edge, which is belonging to the multipolygon. This is eliminating an edge of the multipolygon.
    - If the dragged node and other node are each connected to exactly one edge of the multipolygon.

I will try to make a pull request.
@bhousel 
The geoHasSelfIntesections function expects an array of node entities but the active node as an id.
Is there a specific reason to use an id? The calling functions seem to know the entity of the node anyway, but I'm not sure, if this might be a different version of the node entity.

**Comment by bhousel** [comment]
> The geoHasSelfIntesections function expects an array of node entities but the active node as an id.
> Is there a specific reason to use an id? The calling functions seem to know the entity of the node anyway, but I'm not sure, if this might be a different version of the node entity.

`geoChooseEdge`, `geoHasLineIntersections`, and `geoHasSelfIntersections` all take arrays of nodes to work with, because they need to use `.loc` properties to know where the nodes are.  I don't think there is a specific reason why they take `activeID` instead of `activeNode`, (e.g. performance-wise, I don't think it should matter) but it does make all the testing code simpler to just pass in the `id` of the moving node.

**Comment by bhousel** [comment]
Oh I forgot to mention but... I'm generally ok with iD doing a weird thing in this situation.  I consider this a kind of low-priority bug, because the example case is not something that would really occur on the map, or if it did occur, it's because the user made it themselves, and they can just remove the multipolygon tag.

**Comment by yaugenka** [comment]
@bhousel  it occurs with every polygon or multipolygon consisting of a series of lines.
It causes pain especially when modifying existing polygons of this kind. E.g. when you are fixing administrative boundary and disconnect a line to redraw it or replace with another one, you may get stuck wondering why it won't let you make a closed polygon.

**Comment by bhousel** [comment]
> @bhousel it occurs with every polygon or multipolygon consisting of a series of lines.
> It causes pain especially when modifying existing polygons of this kind

What happens when you remove the `multipolygon` tag?  Does that fix the issue?

**Comment by yaugenka** [comment]
Yes, it happens only with the `multypolygon` tag. But multipolygons are used very often and it is hard to guess this workarround.

**Comment by bhousel** [comment]
> But multipolygons are used very often and it is hard to guess this workarround.

Yes they are used very often, but they are supposed to be non-touching closed rings.

**Comment by yaugenka** [comment]
Here is another example. Assume there is a lake and a wetlend which were initially drawn as two separate geometries but are supposed to run into one another. So you cut out a line from the lake, include a line from the wetlend multypolygon to the lake one, and then try to connect the edges of the broken multypolygon to the new line in order to make the multypolygon closed again. But it won't let you do that.
![line_connect2](https://user-images.githubusercontent.com/14937001/35940784-4c8a138c-0c61-11e8-9af6-7c4a14fd7814.gif)

**Comment by bborkmiller** [comment]
But if you connect the lake to the wetland before you add the shared way to the lake relation--which seems like a more natural order of operations to me--then it works.

**Comment by yaugenka** [comment]
You are right. It happens only when a multipolygon has two unlinked points and one is not lucky with the direction of the lines which he is trying to connect.

**Comment by bhousel** [comment]
> Here is another example. Assume there is a lake and a wetlend which were initially drawn as two separate geometries but are supposed to run into one another. So you cut out a line from the lake, include a line from the wetlend multypolygon to the lake one, and then try to connect the edges of the broken multypolygon to the new line in order to make the multypolygon closed again. But it won't let you do that.

Please remove the multipolygon tag if you are going to do this.
